pkg/cmd/experimental/syncgroups/grouppruner.go

```diff
@@ -0,0 +1,91 @@
+package syncgroups
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+
+	"github.com/openshift/origin/pkg/auth/ldaputil"
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
+	userapi "github.com/openshift/origin/pkg/user/api"
+)
+
+// GroupPruner runs a prune job on Groups
+type GroupPruner interface {
+	Prune() (errors []error)
+}
+
+// LDAPGroupPruner prunes Groups referencing records on an external LDAP server
+type LDAPGroupPruner struct {
+	// Lists all groups to be synced
+	GroupLister interfaces.LDAPGroupLister
+	// Fetches a group and extracts object metainformation and membership list from a group
+	GroupMemberExtractor interfaces.LDAPMemberExtractor
+	// Maps an LDAP group enrty to an OpenShift Group's Name
+	GroupNameMapper interfaces.LDAPGroupNameMapper
+	// Allows the Syncer to search for OpenShift Groups
+	GroupClient client.GroupInterface
+	// Host stores the address:port of the LDAP server
+	Host string
+	// DryRun indicates that no changes should be made.
+	DryRun bool
+
+	// Out is used to provide output while the sync job is happening
+	Out io.Writer
+	Err io.Writer
+}
+
+// Prune allows the LDAPGroupSyncer to be a GroupSyncer
+func (s *LDAPGroupSyncer) Prune() ([]*userapi.Group, []error) {
+	groupsToPrune := []*userapi.Group{}
+	var errors []error
+
+	// determine what to sync
+	glog.V(1).Infof("LDAPGroupPruner listing groups to prune with %v", s.GroupLister)
+	ldapGroupUIDs, err := s.GroupLister.ListGroups()
+	if err != nil {
+		errors = append(errors, err)
+		return nil, errors
+	}
+	glog.V(1).Infof("LDAPGroupPruner will attempt to prune ldapGroupUIDs %v", ldapGroupUIDs)
+
+	for _, ldapGroupUID := range ldapGroupUIDs {
+		glog.V(1).Infof("Checking LDAP group %v", ldapGroupUID)
+
+		// get membership data
+		_, err := s.GroupMemberExtractor.ExtractMembers(ldapGroupUID)
+		if ldaputil.IsNoSuchObjectError(err) || ldaputil.IsEntryNotFoundError(err) {
+			// if the LDAP entry that was previously used to create the group doesn't exist, prune it
+			groupName, err := s.GroupNameMapper.GroupNameFor(ldapGroupUID)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+
+			if !s.DryRun {
+				fmt.Fprintf(s.Out, "group/%s\n", groupName)
+				if err := s.GroupClient.Delete(groupName); err != nil {
+					fmt.Fprintf(s.Err, "Error pruning OpenShift group %q for LDAP group %q: %v.\n", groupName, ldapGroupUID, err)
+					errors = append(errors, err)
+					continue
+				}
+			} else {
+				group, err := s.GroupClient.Get(groupName)
```

Comment:
We don't want to follow the same paradigm of returning the list of groups so that on a dry run the command can work with that?